File: backend/telegram_service.py
import os
import logging
import httpx
from backend.database import get_db_cursor

TELEGRAM_BOT_TOKEN = os.getenv("TELEGRAM_BOT_TOKEN", "")
SITE_URL = os.getenv("SITE_URL", "http://localhost:3000")


def send_message(chat_id: str, text: str) -> bool:
    if not TELEGRAM_BOT_TOKEN or not chat_id:
        return False
    try:
        from backend.proxy_service import get_proxy_url
        proxy_url = get_proxy_url()
        proxies = {"all://": proxy_url} if proxy_url else None
        with httpx.Client(proxies=proxies, timeout=10) as client:
            resp = client.post(
                f"https://api.telegram.org/bot{TELEGRAM_BOT_TOKEN}/sendMessage",
                json={"chat_id": chat_id, "text": text, "parse_mode": "HTML"},
            )
        return resp.status_code == 200
    except Exception as e:
        logging.warning("Telegram send_message failed: %s", e)
        return False


def get_chat_id_by_related(role: str, related_id: int) -> str:
    try:
        with get_db_cursor() as cur:
            cur.execute(
                "SELECT telegram_chat_id FROM users WHERE role = %s AND related_id = %s",
                (role, related_id),
            )
            row = cur.fetchone()
            return row["telegram_chat_id"] if row else None
    except Exception:
        return None


def _also_web_push(role: str, related_id: int, text: str):
    # Web Push mirrors every Telegram notification so users without the bot
    # still get pushes in the browser/PWA. Best-effort, never blocks.
    try:
        from backend import push_service
        push_service.notify_role(role, related_id, text)
    except Exception:
        pass


def notify_needy(needy_id: int, text: str):
    chat_id = get_chat_id_by_related("needy", needy_id)
    if chat_id:
        send_message(chat_id, text)
    _also_web_push("needy", needy_id, text)


def notify_shop(shop_id: int, text: str):
    chat_id = get_chat_id_by_related("shop", shop_id)
    if chat_id:
        send_message(chat_id, text)
    _also_web_push("shop", shop_id, text)


def notify_volunteer(volunteer_id: int, text: str):
    chat_id = get_chat_id_by_related("volunteer", volunteer_id)
    if chat_id:
        send_message(chat_id, text)
    _also_web_push("volunteer", volunteer_id, text)

File: backend-java/src/main/java/ru/savefood/telegram/TelegramService.java
package ru.savefood.telegram;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.OutputStream;
import java.net.InetSocketAddress;
import java.net.Proxy;
import java.net.URI;
import java.net.URL;
import java.net.HttpURLConnection;
import java.nio.charset.StandardCharsets;
import java.util.Map;
import java.util.logging.Logger;
import ru.savefood.proxy.ProxyService;
import ru.savefood.push.PushDispatchService;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.stereotype.Service;
@Service
public class TelegramService {
    private static final Logger log = Logger.getLogger(TelegramService.class.getName());
    private final ObjectMapper mapper = new ObjectMapper();
    private final JdbcTemplate jdbc;
    private final ProxyService proxyService;
    private final PushDispatchService push;
    private final String botToken;
    public TelegramService(JdbcTemplate jdbc, ProxyService proxyService, PushDispatchService push,
                           @Value("${savefood.oauth.telegram-bot-token:}") String botToken) {
        this.jdbc = jdbc;
        this.proxyService = proxyService;
        this.push = push;
        this.botToken = botToken == null ? "" : botToken;
    }
    /** Low-level send. Returns true on HTTP 200, false (never throws) otherwise. */
    public boolean sendMessage(String chatId, String text) {
        if (botToken.isEmpty() || chatId == null || chatId.isEmpty()) {
            return false;
        }
        try {
            byte[] body = mapper.writeValueAsBytes(Map.of(
                "chat_id", chatId, "text", text, "parse_mode", "HTML"));
            URL url = URI.create(
                "https://api.telegram.org/bot" + botToken + "/sendMessage").toURL();
            HttpURLConnection conn = (HttpURLConnection) url.openConnection(proxy());
            conn.setRequestMethod("POST");
            conn.setConnectTimeout(10_000);
            conn.setReadTimeout(10_000);
            conn.setDoOutput(true);
            conn.setRequestProperty("Content-Type", "application/json");
            try (OutputStream os = conn.getOutputStream()) {
                os.write(body);
            }
            int status = conn.getResponseCode();
            conn.disconnect();
            return status == 200;
        } catch (Exception e) {
            log.warning("Telegram send_message failed: " + e.getMessage());
            return false;
        }
    }
    public String getChatIdByRelated(String role, int relatedId) {
        try {
            return jdbc.query(
                "SELECT telegram_chat_id FROM users WHERE role = ? AND related_id = ?",
                rs -> rs.next() ? rs.getString("telegram_chat_id") : null, role, relatedId);
        } catch (RuntimeException e) {
            return null;
        }
    }
    public void notifyNeedy(int needyId, String text) {
        notify("needy", needyId, text);
    }
    public void notifyShop(int shopId, String text) {
        notify("shop", shopId, text);
    }
    public void notifyVolunteer(int volunteerId, String text) {
        notify("volunteer", volunteerId, text);
    }
    private void notify(String role, int relatedId, String text) {
        String chatId = getChatIdByRelated(role, relatedId);
        if (chatId != null && !chatId.isEmpty()) {
            sendMessage(chatId, text);
        }
        try {
            push.notifyRole(role, relatedId, text, "/");
        } catch (Exception ignore) {
        }
    }
    private Proxy proxy() {
        String url = proxyService.getProxyUrl();
        if (url == null || !url.startsWith("socks5://")) {
            return Proxy.NO_PROXY;
        }
        String hostPort = url.substring("socks5://".length());
        int colon = hostPort.lastIndexOf(':');
        if (colon <= 0) {
            return Proxy.NO_PROXY;
        }
        String host = hostPort.substring(0, colon);
        int port = Integer.parseInt(hostPort.substring(colon + 1));
        return new Proxy(Proxy.Type.SOCKS, new InetSocketAddress(host, port));
    }
}
